source/uwp/Renderer/lib/XamlBuilder.cpp

```diff
@@ -2768,6 +2750,120 @@ AdaptiveNamespaceStart
+        THROW_IF_FAILED(mediaElement->put_AutoPlay(false));
+
+        ComPtr<IMediaElement2> mediaElement2;
+        THROW_IF_FAILED(mediaElement.As(&mediaElement2));
+        THROW_IF_FAILED(mediaElement2->put_AreTransportControlsEnabled(true));
+
+        ComPtr<IUIElement> mediaUIElement;
+        THROW_IF_FAILED(mediaElement.As(&mediaUIElement));
+
+        if (posterImage == nullptr)
+        {
+            // If there's no poster, just return the media element
+            mediaElement.CopyTo(mediaControl);
+        }
+        else
+        {
+            // Put the poster image in a container with a play button
+            ComPtr<IUIElement> posterContainer;
+            CreatePosterContainerWithPlayButton(posterImage.Get(), renderArgs, hostConfig.Get(), &posterContainer);
+
+            // Set the poster on the media element
+            ComPtr<IImageSource> posterImageSource;
+            THROW_IF_FAILED(posterImage->get_Source(&posterImageSource));
+            THROW_IF_FAILED(mediaElement->put_PosterSource(posterImageSource.Get()));
+
+            // Make the media element collapsed until the user clicks
+            THROW_IF_FAILED(mediaUIElement->put_Visibility(Visibility_Collapsed));
+
+            // Create a panel to hold the poster and the media element
+            ComPtr<IStackPanel> mediaStackPanel = XamlHelpers::CreateXamlClass<IStackPanel>(HStringReference(RuntimeClass_Windows_UI_Xaml_Controls_StackPanel));
+            ComPtr<IPanel> mediaPanel;
+            THROW_IF_FAILED(mediaStackPanel.As(&mediaPanel));
+
+            XamlHelpers::AppendXamlElementToPanel(posterContainer.Get(), mediaPanel.Get());
+            XamlHelpers::AppendXamlElementToPanel(mediaElement.Get(), mediaPanel.Get());
+
+            // Wrap in touch target
+            ComPtr<IUIElement> mediaPanelAsUIElement;
+            THROW_IF_FAILED(mediaPanel.As(&mediaPanelAsUIElement));
+
+            ComPtr<IUIElement> touchTargetUIElement;
+            WrapInTouchTarget(adaptiveCardElement, mediaPanelAsUIElement.Get(), nullptr, renderContext, true, &touchTargetUIElement);
+
+            ComPtr<IButtonBase> touchTargetAsButtonBase;
+            THROW_IF_FAILED(touchTargetUIElement.As(&touchTargetAsButtonBase));
+
+            // When the user clicks: hide the poster, show the media element, play the media
+            EventRegistrationToken clickToken;
+            THROW_IF_FAILED(touchTargetAsButtonBase->add_Click(Callback<IRoutedEventHandler>([posterContainer, mediaElement](IInspectable* /*sender*/, IRoutedEventArgs* /*args*/) -> HRESULT
+            {
+                THROW_IF_FAILED(posterContainer->put_Visibility(Visibility_Collapsed));
+
+                ComPtr<IUIElement> mediaAsUIElement;
+                THROW_IF_FAILED(mediaElement.As(&mediaAsUIElement));
+                THROW_IF_FAILED(mediaAsUIElement->put_Visibility(Visibility_Visible));
+
+                THROW_IF_FAILED(mediaElement->Play());
+
```

Comment:
Is throwing from an EventHandler the right thing to do?

Reply:
Nope. Switched to RETURN_IF_FAILED.

---
In reply to: [190370228](https://github.com/Microsoft/AdaptiveCards/pull/1490#discussion_r190370228) [](ancestors = 190370228)